### ISSUE TITLE
Ping/Latency substantially affects client or and server performance.

### ISSUE THREAD

**Issue by SemixOfficial** [HaodongMo/ArcCW#55]
Hello, so few hours ago I was just messing around with some things and I noticed that physical bullets are not lag compensated, being a total nerd I have decided to lag compensate them myself and rewrite ArcCW physbullet code to make projectiles fully lag compensated and supported, whilst doing so I have ran into a very interesting issue however, I have noticed that regardless of what I did player's latency to the server has a huge impact on framerate and because I couldn't find and fix this issue myself I have decided to submit a issue here on the github, running a powerful Ryzen 5 5600X CPU and RTX 2060 Super OC GPU I can usually stay on stable 143 fps (limited by RTSS) regardless of what I do or where I play, yet with ArcCW weapon a round trip time of 400ms results in massive loads and as shown in the video below, I struggle to get stable or even decent frametimes, and on round trip time of 800ms the game becomes completely unplayable as the client experiences massive rubber banding and stutter, sure no one is going to play or have nice time playing with round trip time of 400ms or even 200ms but it still shows that somewhere in ArcCW a bit of code has a lot of performance issues as latency increases.

I feel like this is a really big issue and whilst I don't know whether it's the server or client that is experiencing massive loads as latency goes up due to testing on a local server I still believe that this is an issue that needs a lot of attention and has to be fixed as quickly as possible as it can very negatively affect people with weaker PC setups or high latency.

Here is the link to my video showcasing the issue: https://youtu.be/CQCtdfqYGs8

**Comment by SemixOfficial** [comment]
Hello, after nearly one year I have made another investigation into this problem and I have managed to pin-point the smaller portion of this problem to come from SWEP:GetViewModelPosition located [here](https://github.com/HaodongMo/ArcCW/blob/master/lua/weapons/arccw_base/cl_viewmodel.lua#L203).

This function unfortunately suffers from multitude of issues, the excessive lag being the main one, so let's begin with it.

GetViewModelPosition is a predicted hook which means that in the client realm this hook will be called multiple times in order to smooth out the experience as the client simulates multiple times at different intervals, interpolating the results, as the client's latency to the server increases and predicted hooks need to be called more often, GetViewModelPosition alone isn't all that expensive, during my testing, with zero latency it had a stable average runtime of 240uS, generating about 3.9 kilobytes of garbage every call, the problem comes when your latency starts increasing, with round trip time of 200msec this becomes rather unstable runtime of anywhere between 1500-2700uS, generating 50.7kb of garbage every call, so on top of low frame rate due to to the cost of calling the function so many times, the amount of garbage generated will trigger a full garbage collection cycle almost every 0.2-0.4 seconds, this makes the game completely unplayable at any latency that's above 100-150 milliseconds, which is a large issue in my opinion, now do keep in mind that I am on a local server with no clutter or other addons that could in any noticeable way affect my test results, anyways here are some interesting screenshots to take a look at.

RTSS Frametime Graph (ArcCW, Zero latency)
![gmod_j1gb6lv3pX](https://user-images.githubusercontent.com/38821091/159753085-49d2edaf-874c-4b67-8047-fe4a3ede805d.png)

RTSS Frametime Graph (ArcCW, 200msec latency)
![gmod_FFNtMnzPl0](https://user-images.githubusercontent.com/38821091/159753271-ea9b2460-32d2-40d9-a676-46606ec9b787.png)

RTSS Frametime Graph (Customisable Weaponry 2.0, 200msec latency)
![image](https://user-images.githubusercontent.com/38821091/159753539-7df29beb-dbaa-46e3-86dd-e972e22b850f.png)

RTSS Frametime Graph (Customisable Weaponry 2.0, 1000msec latency)
![image](https://user-images.githubusercontent.com/38821091/159753770-2188d8b6-7da0-450a-85b3-5ecd86247141.png)


FProfiler test results (Focused on GetViewModelPosition only)
![gmod_svafAT1A8h](https://user-images.githubusercontent.com/38821091/159752929-560fc5a6-af7a-4943-8df0-eb7f3bcd250c.png)

Frametime Instabilities at 200msec (ArcCW)
![image](https://user-images.githubusercontent.com/38821091/159755807-0ad71af5-1adc-4947-8b75-1993167ad2ee.png)

All of the tests were done on a local P2P server with 7empty slots (8 in total) using arccw_ak47 from Counter-Strike+ pack with no attachments and all ArcCW settings set to default values, the only modified settings were: 

Server settings: `sv_cheats 1;net_fakelag 100;sv_maxrate 0`
Client settings: `cl_SetupAllBones 0;cl_threaded_bone_setup 1;mat_queue_mode 2;gmod_mcore_test 1;cl_predictionlist 1;net_graph 4`

To conclude this, I have tried fixing this issue myself but after few hours of testing I came to the conclusion that not being one of the developers I don't understand majority of the code for stuff like GetBuff or why such system exists, I have vague idea as to what most of it does but when it comes to fixing issues having vague idea about something is usually not the best approach, therefore I am leaving this issue to the original developers here on GitHub to hopefully fix up the code as they know their weapon pack the best. (obviously)

That being said GetViewModelPosition is not the only issue, compared to other weapon packs like M9K, CW 2.0, FA:S, ArcCW seems to have a lot of problems with prediction and or predicted hooks, I understand that this weapon pack is lot larger and has many more features than the above mentioned ones, and I have nothing but respect for the developers who put their time and hard work into this, but I do believe that some bits of the code need to be re-worked.

**Comment by Fesiug** [comment]
Development efforts are being nowadays being pushed towards [ARC-9](https://github.com/HaodongMo/ARC-9/), so if you would like to do profiling and stuff for that, that would help

**Comment by TheOnly8Z** [comment]
@SemixOfficial I made a preliminary optimization to reduce GetBuff calls on non-changing values in the optimization branch. Please try it out and see if it is an improvement.

https://github.com/HaodongMo/ArcCW/tree/optimization

**Comment by SemixOfficial** [comment]
> Development efforts are being nowadays being pushed towards [ARC-9](https://github.com/HaodongMo/ARC-9/), so if you would like to do profiling and stuff for that, that would help

I have just yesterday written some basic parts of my own weapon pack, I would be happy to join ARC-9 development team if possible.

**Comment by TheOnly8Z** [comment]
> The improvement is still significant, and given the fact that the frametime graph is almost stable on 200msec latency I deem the runtime optimization to be good enough for now, at 200msec latency users might have larger concerns than ~0.01ms frametime instabilities, what needs to be focused on now is the garbage collection, even with 100ms latency, GetViewModelPosition still seems to be triggering manual garbage collection, as can be seen here with the sudden spike ![image](https://user-images.githubusercontent.com/38821091/159987680-9ce89c9a-34ea-4a65-b2e7-fa7b643ecd53.png) and here ![image](https://user-images.githubusercontent.com/38821091/159989055-b590c330-4d90-42b4-876c-baee4377f27a.png)

Can you show the code you use to calculate the render time and garbage size?

**Comment by SemixOfficial** [comment]
I create two trash/placeholder variables in the file's scope like so 
![image](https://user-images.githubusercontent.com/38821091/160032855-85a83aa1-84b8-44b3-8f25-8767b8c80ffa.png)
then in PreDrawiewModel I assign them values, one being SysTime() used for render time, and the second being amount of allocated dynamic lua memory
![image](https://user-images.githubusercontent.com/38821091/160032898-682df44b-7361-406d-af4a-a18f67827f6d.png)
then in PostDrawViewModel I simply print the difference between the current values and the values before. 
![image](https://user-images.githubusercontent.com/38821091/160033253-2e70ad92-a515-42b8-bc00-1c44f2277e6e.png)
(Note that the SysTime delta is multiplied by 1,000,000 because one second has 1,000,00 microseconds (uS)

Do note that this is not perfect way to measure the total cost of GetViewModelPosition, as we are measuring it per call and not per frame, but I see that as tolerable, because we only need to know how much the function takes per call to see if we are optimizing it or making improvements, since the final result will be your average cost per call times the amount of times the function has been called, the most reliable way to measure the total cost of GetViewModelPosition I can think of would be to create two variables in global scope, print and reset their values in FrameUpdatePreEntityThink, then for every timeGetViewModelPosition is called measure the time and garbage, add it to the placeholder variables and see the results on next frame, this should tell us the total cost of GetViewModelPosition as it would include additional prediction calls from the engine, but again, It's not really necessary info.

**Comment by TheOnly8Z** [comment]
@SemixOfficial I've pushed changes to the branch that should noticeably decrease garbage generation. Let me know how it goes.

**Comment by SemixOfficial** [comment]
I have tested the latest changes, in the optimization branch, the improvements are good, although the weapons still trigger garbage collection occasionally on higher latency, they don't trigger full cycle, but only smaller steps, which are tolerable and don't lag the game for full second or so, there is a large problem that I have noticed tho, that being view model bob, sway and iron sights not working at all now and view model appears in weird position, this is most likely caused by swapping most of add, substract, multiply and other operations from regular arithmetic to function calls like `VMPos:Add()`, etc.... as these functions modify the Vector or Angle itself rather than creating a new copy or changing just the localized reference.

Here are some in-game screenshots:
![gmod_rtxR443zik](https://user-images.githubusercontent.com/38821091/160183975-22d2de62-2ce7-4d47-afeb-0e129f3923f3.png)
![gmod_ldhTRPZP5P](https://user-images.githubusercontent.com/38821091/160183984-142be929-43a2-4c63-a0b0-fbd062295e52.png)

as far as today's profiling results go, GetViewModelPosition still appears to be amongst one of the largest bottlenecks, but table.Add and GetBuff_Override functions seem to be catching up, both spotting awful cost of 3500-4000uS per frame.

anyways, the important part is that the weapon pack is very usable now even with high latency, which is what matters the most, therefore I deem the issue fixed and will probably close this ticket, further improvements could still be made but with development efforts being pushed towards ARC-9 It doesn't make a whole lot of sense to rewrite every single expensive bit of ArcCW.
